File: 023_merge_k_sorted_list/recur.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        if not l1:
            return l2
        if not l2:
            return l1

        if l1.val < l2.val:
            l1.next = self.mergeTwoLists(l1.next, l2)
            return l1
        else:
            l2.next = self.mergeTwoLists(l1, l2.next)
            return l2


s = Solution()
list1 = [1, 6, 7, 8]
list2 = [2, 3, 5, 8, 9]
l1 = ListNode(list1[0])
cur = l1
for i in range(1, len(list1)):
    cur.next = ListNode(list1[i])
    cur = cur.next
l2 = ListNode(list2[0])
cur = l2
for i in range(1, len(list2)):
    cur.next = ListNode(list2[i])
    cur = cur.next
cur = s.mergeTwoLists(l1, l2)
while cur:
    print(cur.val)
    cur = cur.next

File: 023_merge_k_sorted_list/recur.c
#include <stdlib.h>
#include <string.h>
#include <assert.h>
struct ListNode {
    int val;
    struct ListNode *next;
};
 


struct ListNode* mergeTwoLists(struct ListNode* list1, struct ListNode* list2){
    if (list1 == NULL)
        return list2;
    if (list2 == NULL)
        return list1;
  
    if (list1->val < list2->val) {
        list1->next = merge(list1->next, list2);
        return list1;
    } else {
        list2->next = merge(list1, list2->next);
        return list2;
    }

}
struct ListNode* mergeKLists(struct ListNode** lists, int listsSize){
    struct ListNode  *new_head = NULL;
    
    for(int i=0;i<listsSize;i++){
        new_head = mergeTwoLists(new_head,lists[i]);
    }
    
    return new_head;
}